File: src/main/java/io/koju/autopos/trade/purchase/domain/PurchaseLine.java
package io.koju.autopos.trade.purchase.domain;

import com.fasterxml.jackson.annotation.JsonBackReference;
import io.koju.autopos.trade.domain.LineItem;
import lombok.Getter;
import lombok.Setter;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.validation.constraints.NotNull;

import static javax.persistence.GenerationType.SEQUENCE;

@Entity
@Table(name = "purchase_line")
@Getter
@Setter
public class PurchaseLine extends LineItem {

    private static final String ID_SEQ = "purchase_line_id_seq";

    @Id
    @SequenceGenerator(name = ID_SEQ, sequenceName = ID_SEQ, allocationSize = 1)
    @GeneratedValue(strategy = SEQUENCE, generator = ID_SEQ)
    private Long id;

    @NotNull
    @ManyToOne(optional = false)
    @JoinColumn(name = "purchase_id")
    @JsonBackReference
    private Purchase purchase;

    @Column(name = "buyer")
    private String buyer;

    @Override
    public Long getId() {
        return id;
    }
}

File: src/main/scala/io/koju/autopos/trade/purchase/domain/PurchaseLine.scala
package io.koju.autopos.trade.purchase.domain

import java.lang.Long
import java.math.BigDecimal
import javax.persistence.GenerationType.SEQUENCE
import javax.persistence._
import javax.validation.constraints.{Min, NotNull}

import com.fasterxml.jackson.annotation.JsonBackReference
import io.koju.autopos.trade.domain.LineItem

import scala.beans.BeanProperty

@Entity
@Table(name = "purchase_line")
class PurchaseLine extends LineItem {

  @Id
  @SequenceGenerator(name = "purchase_line_id_seq", sequenceName = "purchase_line_id_seq", allocationSize = 1)
  @GeneratedValue(strategy = SEQUENCE, generator = "purchase_line_id_seq")
  @BeanProperty
  var id: Long = _

  @NotNull
  @Min(value = 0)
  @Column(name = "rate", precision = 10, scale = 2, nullable = false)
  @BeanProperty
  var rate: BigDecimal = _

  @NotNull
  @ManyToOne(optional = false)
  @JoinColumn(name = "purchase_id")
  @JsonBackReference
  @BeanProperty
  var purchase: Purchase = _
}